File: apps/burner-mac/Tests/FlagshipBurnerTests/VMLifecycleTests.swift
import XCTest
@testable import FlagshipBurnerCore

/// Every transition of the pure VM state machine, including the
/// install→first-boot ISO-detach seam and the sealed
/// "waiting for you to unlock" state.
final class VMLifecycleTests: XCTestCase {

    private func sealed(_ state: VMState = .created,
                        clock: @escaping @Sendable () -> Date = { Date(timeIntervalSince1970: 0) }) -> VMLifecycle {
        VMLifecycle(state: state, sealedAtBoot: true, clock: clock)
    }

    private func unsealed(_ state: VMState = .created) -> VMLifecycle {
        VMLifecycle(state: state, sealedAtBoot: false, clock: { Date(timeIntervalSince1970: 0) })
    }

    // MARK: - Install phase

    func testStartInstallAttachesISOAndStarts() throws {
        var lc = sealed()
        let effects = try lc.handle(.startInstall)
        XCTAssertEqual(lc.state, .installing)
        XCTAssertEqual(effects, [.attachInstallerISO, .startVirtualMachine])
    }

    func testInstallSucceededDetachesTheISO() throws {
        // The seam the spec calls out: after the unattended install the ISO
        // comes OFF, so every later boot is from the guest's own disk.
        var lc = sealed(.installing)
        let effects = try lc.handle(.installSucceeded)
        XCTAssertEqual(lc.state, .installed)
        XCTAssertEqual(effects, [.detachInstallerISO])
    }

    func testInstallFailureStopsAndDetaches() throws {
        var lc = sealed(.installing)
        let effects = try lc.handle(.installFailed("guest error"))
        XCTAssertEqual(lc.state, .failed(VMFailure(phase: .install, reason: "guest error")))
        XCTAssertEqual(effects, [.stopVirtualMachine, .detachInstallerISO])
    }

    func testFailedInstallIsRetryable() throws {
        var lc = sealed(.failed(VMFailure(phase: .install, reason: "x")))
        let effects = try lc.handle(.startInstall)
        XCTAssertEqual(lc.state, .installing)
        XCTAssertEqual(effects, [.attachInstallerISO, .startVirtualMachine])
    }

    // MARK: - Boot: the sealed state

    func testEncryptedGuestBootsIntoAwaitingPhoneUnlock() throws {
        var lc = sealed(.installed)
        let effects = try lc.handle(.powerOn)
        XCTAssertEqual(lc.state, .awaitingPhoneUnlock)
        XCTAssertEqual(effects, [.startVirtualMachine])
    }

    func testUnencryptedGuestBootsStraightToRunning() throws {
        var lc = unsealed(.installed)
        let effects = try lc.handle(.powerOn)
        XCTAssertEqual(lc.state, .running)
        XCTAssertEqual(effects, [.startVirtualMachine])
    }

    func testPhoneUnlockCompletesTheBoot() throws {
        var lc = sealed(.awaitingPhoneUnlock)
        let effects = try lc.handle(.guestUnlocked)
        XCTAssertEqual(lc.state, .running)
        XCTAssertEqual(effects, [])
    }

    func testSealedGuestCanBePoweredOffWhileWaiting() throws {
        // The VM boots with the host but stays sealed — the owner may still
        // shut it down without ever unlocking.
        var lc = sealed(.awaitingPhoneUnlock)
        let effects = try lc.handle(.powerOff)
        XCTAssertEqual(lc.state, .stopped)
        XCTAssertEqual(effects, [.stopVirtualMachine])
    }

    // MARK: - Run / stop / restart

    func testRunningPowersOffToStopped() throws {
        var lc = sealed(.running)
        let effects = try lc.handle(.powerOff)
        XCTAssertEqual(lc.state, .stopped)
        XCTAssertEqual(effects, [.stopVirtualMachine])
    }

    func testStoppedRebootsThroughTheSealedState() throws {
        var lc = sealed(.stopped)
        let effects = try lc.handle(.powerOn)
        XCTAssertEqual(lc.state, .awaitingPhoneUnlock)
        XCTAssertEqual(effects, [.startVirtualMachine])
    }

    func testRuntimeFailureFromRunning() throws {
        var lc = sealed(.running)
        _ = try lc.handle(.runtimeFailed("crashed"))
        XCTAssertEqual(lc.state, .failed(VMFailure(phase: .run, reason: "crashed")))
    }

    func testRuntimeFailureWhileAwaitingUnlock() throws {
        var lc = sealed(.awaitingPhoneUnlock)
        _ = try lc.handle(.runtimeFailed("died"))
        XCTAssertEqual(lc.state, .failed(VMFailure(phase: .run, reason: "died")))
    }

    func testRunFailureIsRestartable() throws {
        var lc = sealed(.failed(VMFailure(phase: .run, reason: "x")))
        let effects = try lc.handle(.powerOn)
        XCTAssertEqual(lc.state, .awaitingPhoneUnlock)
        XCTAssertEqual(effects, [.startVirtualMachine])
    }

    // MARK: - Invalid transitions are loud, not swallowed

    func testInvalidTransitionsThrow() {
        let cases: [(VMState, VMEvent)] = [
            (.created, .powerOn),               // nothing installed yet
            (.created, .installSucceeded),
            (.installing, .powerOn),            // mid-install
            (.installing, .guestUnlocked),
            (.installed, .startInstall),        // already installed
            (.installed, .guestUnlocked),       // not booted
            (.running, .powerOn),               // already up
            (.running, .guestUnlocked),
            (.running, .startInstall),
            (.stopped, .guestUnlocked),
            (.awaitingPhoneUnlock, .powerOn),
            (.failed(VMFailure(phase: .install, reason: "x")), .powerOn), // must retry install
            (.failed(VMFailure(phase: .run, reason: "x")), .startInstall),
        ]
        for (state, event) in cases {
            var lc = sealed(state)
            XCTAssertThrowsError(try lc.handle(event), "\(state) + \(event) should throw") { err in
                XCTAssertEqual(err as? VMLifecycleError,
                               .invalidTransition(from: state, on: event))
            }
            XCTAssertEqual(lc.state, state, "state must not change on a rejected event")
        }
    }

    // MARK: - Install-stop interpretation (reboot / poweroff / never-booted)

    func testCleanInstallStopAfterPlausibleDurationIsInstalled() {
        // A real unattended install runs for minutes; a clean stop once that
        // much time has passed is a genuine completion — whether the finished
        // installer POWERED OFF (the shipping preseed) or REBOOTED (a base image
        // whose finish-install reboots). VZ surfaces both as a clean
        // guestDidStop, so the verdict is duration-based and covers both.
        let start = Date(timeIntervalSince1970: 0)
        let poweroff = start.addingTimeInterval(600) // 10 min later
        XCTAssertEqual(VMLifecycle.verdictForCleanInstallStop(installStartedAt: start, now: poweroff),
                       .installed)
        // Reboot-at-end path: same signal (clean stop), also well past the floor.
        let reboot = start.addingTimeInterval(VMLifecycle.minPlausibleInstallDuration + 1)
        XCTAssertEqual(VMLifecycle.verdictForCleanInstallStop(installStartedAt: start, now: reboot),
                       .installed)
    }

    func testFastCleanStopIsFailureNotInstalled() {
        // The Phase-0 hardware finding: an amd64 base image on an Apple-silicon
        // host makes VZ stop the guest ~0.3s after start with error=nil. That is
        // NOT a finished install — reading it as success boots an empty disk.
        let start = Date(timeIntervalSince1970: 0)
        let instantStop = start.addingTimeInterval(0.3)
        guard case .failedTooFast(let elapsed) =
                VMLifecycle.verdictForCleanInstallStop(installStartedAt: start, now: instantStop) else {
            return XCTFail("a 0.3s clean stop must be failedTooFast, not installed")
        }
        XCTAssertEqual(elapsed, 0.3, accuracy: 0.001)
    }

    func testInstallStopVerdictBoundary() {
        let start = Date(timeIntervalSince1970: 0)
        // Exactly at the floor counts as installed (inclusive).
        let atFloor = start.addingTimeInterval(VMLifecycle.minPlausibleInstallDuration)
        XCTAssertEqual(VMLifecycle.verdictForCleanInstallStop(installStartedAt: start, now: atFloor),
                       .installed)
        let justUnder = start.addingTimeInterval(VMLifecycle.minPlausibleInstallDuration - 0.5)
        guard case .failedTooFast = VMLifecycle.verdictForCleanInstallStop(installStartedAt: start, now: justUnder) else {
            return XCTFail("just under the floor must be failedTooFast")
        }
    }

    func testComingUpStallAdvisory() {
        // A sealed guest awaiting unlock should surface a "taking longer than
        // expected" advisory once it has waited past the stall threshold — but
        // only in the awaiting-unlock state, and not before the threshold.
        let below = VMLifecycle.comingUpStallThreshold - 1
        let above = VMLifecycle.comingUpStallThreshold + 1
        XCTAssertFalse(VMLifecycle.comingUpIsStalled(state: .awaitingPhoneUnlock, elapsed: below))
        XCTAssertTrue(VMLifecycle.comingUpIsStalled(state: .awaitingPhoneUnlock, elapsed: above))
        // Exactly at the threshold trips (inclusive).
        XCTAssertTrue(VMLifecycle.comingUpIsStalled(state: .awaitingPhoneUnlock,
                                                    elapsed: VMLifecycle.comingUpStallThreshold))
        // Never fires outside the awaiting-unlock state, no matter how long.
        XCTAssertFalse(VMLifecycle.comingUpIsStalled(state: .running, elapsed: above))
        XCTAssertFalse(VMLifecycle.comingUpIsStalled(state: .installing, elapsed: above))
    }

    // MARK: - Injectable clock

    func testStateTimestampsComeFromTheInjectedClock() throws {
        let times: [Date] = [Date(timeIntervalSince1970: 100),
                             Date(timeIntervalSince1970: 200),
                             Date(timeIntervalSince1970: 300)]
        let box = ClockBox(times: times)
        var lc = VMLifecycle(sealedAtBoot: true, clock: { box.next() })
        XCTAssertEqual(lc.stateChangedAt, times[0])
        try lc.handle(.startInstall)
        XCTAssertEqual(lc.stateChangedAt, times[1])
        try lc.handle(.installSucceeded)
        XCTAssertEqual(lc.stateChangedAt, times[2])
    }

    private final class ClockBox: @unchecked Sendable {
        private var times: [Date]
        private let lock = NSLock()
        init(times: [Date]) { self.times = times }
        func next() -> Date {
            lock.lock(); defer { lock.unlock() }
            return times.isEmpty ? Date.distantFuture : times.removeFirst()
        }
    }
}

File: apps/builder-windows/tests/VMLifecycleTests.cs
using System;
using System.Collections.Generic;
using Xunit;
using Flagship.Builder.VM;

namespace Flagship.Builder.Tests;

/// <summary>
/// Direct port of apps/builder-mac VMLifecycleTests.swift: every transition of
/// the pure VM state machine, the install→first-boot ISO-detach seam, the
/// sealed "waiting for you to unlock" state, the injected clock, plus the
/// Windows-side tests for the duration-gated install verdict. The full
/// transition table is ALSO pinned by VMCoreVectorTests (shared vectors);
/// these keep the intent readable test-by-test.
/// </summary>
public class VMLifecycleTests
{
    private static VMLifecycle Sealed(VMState? state = null)
        => new(sealedAtBoot: true, state, () => DateTimeOffset.UnixEpoch);

    private static VMLifecycle Unsealed(VMState? state = null)
        => new(sealedAtBoot: false, state, () => DateTimeOffset.UnixEpoch);

    // ---- Install phase ----

    [Fact]
    public void StartInstallAttachesISOAndStarts()
    {
        var lc = Sealed();
        var effects = lc.Handle(VMEvent.StartInstall);
        Assert.Equal(VMState.Installing, lc.State);
        Assert.Equal(new[] { VMEffect.AttachInstallerISO, VMEffect.StartVirtualMachine }, effects);
    }

    [Fact]
    public void InstallSucceededDetachesTheISO()
    {
        // The seam the spec calls out: after the unattended install the ISO
        // comes OFF, so every later boot is from the guest's own disk.
        var lc = Sealed(VMState.Installing);
        var effects = lc.Handle(VMEvent.InstallSucceeded);
        Assert.Equal(VMState.Installed, lc.State);
        Assert.Equal(new[] { VMEffect.DetachInstallerISO }, effects);
    }

    [Fact]
    public void InstallFailureStopsAndDetaches()
    {
        var lc = Sealed(VMState.Installing);
        var effects = lc.Handle(VMEvent.InstallFailed("guest error"));
        Assert.Equal(VMState.Failed(VMFailurePhase.Install, "guest error"), lc.State);
        Assert.Equal(new[] { VMEffect.StopVirtualMachine, VMEffect.DetachInstallerISO }, effects);
    }

    [Fact]
    public void FailedInstallIsRetryable()
    {
        var lc = Sealed(VMState.Failed(VMFailurePhase.Install, "x"));
        var effects = lc.Handle(VMEvent.StartInstall);
        Assert.Equal(VMState.Installing, lc.State);
        Assert.Equal(new[] { VMEffect.AttachInstallerISO, VMEffect.StartVirtualMachine }, effects);
    }

    // ---- Boot: the sealed state ----

    [Fact]
    public void EncryptedGuestBootsIntoAwaitingPhoneUnlock()
    {
        var lc = Sealed(VMState.Installed);
        var effects = lc.Handle(VMEvent.PowerOn);
        Assert.Equal(VMState.AwaitingPhoneUnlock, lc.State);
        Assert.Equal(new[] { VMEffect.StartVirtualMachine }, effects);
    }

    [Fact]
    public void UnencryptedGuestBootsStraightToRunning()
    {
        var lc = Unsealed(VMState.Installed);
        var effects = lc.Handle(VMEvent.PowerOn);
        Assert.Equal(VMState.Running, lc.State);
        Assert.Equal(new[] { VMEffect.StartVirtualMachine }, effects);
    }

    [Fact]
    public void PhoneUnlockCompletesTheBoot()
    {
        var lc = Sealed(VMState.AwaitingPhoneUnlock);
        var effects = lc.Handle(VMEvent.GuestUnlocked);
        Assert.Equal(VMState.Running, lc.State);
        Assert.Empty(effects);
    }

    [Fact]
    public void SealedGuestCanBePoweredOffWhileWaiting()
    {
        // The VM boots with the host but stays sealed — the owner may still
        // shut it down without ever unlocking.
        var lc = Sealed(VMState.AwaitingPhoneUnlock);
        var effects = lc.Handle(VMEvent.PowerOff);
        Assert.Equal(VMState.Stopped, lc.State);
        Assert.Equal(new[] { VMEffect.StopVirtualMachine }, effects);
    }

    // ---- Run / stop / restart ----

    [Fact]
    public void RunningPowersOffToStopped()
    {
        var lc = Sealed(VMState.Running);
        var effects = lc.Handle(VMEvent.PowerOff);
        Assert.Equal(VMState.Stopped, lc.State);
        Assert.Equal(new[] { VMEffect.StopVirtualMachine }, effects);
    }

    [Fact]
    public void StoppedRebootsThroughTheSealedState()
    {
        var lc = Sealed(VMState.Stopped);
        var effects = lc.Handle(VMEvent.PowerOn);
        Assert.Equal(VMState.AwaitingPhoneUnlock, lc.State);
        Assert.Equal(new[] { VMEffect.StartVirtualMachine }, effects);
    }

    [Fact]
    public void RuntimeFailureFromRunning()
    {
        var lc = Sealed(VMState.Running);
        lc.Handle(VMEvent.RuntimeFailed("crashed"));
        Assert.Equal(VMState.Failed(VMFailurePhase.Run, "crashed"), lc.State);
    }

    [Fact]
    public void RuntimeFailureWhileAwaitingUnlock()
    {
        var lc = Sealed(VMState.AwaitingPhoneUnlock);
        lc.Handle(VMEvent.RuntimeFailed("died"));
        Assert.Equal(VMState.Failed(VMFailurePhase.Run, "died"), lc.State);
    }

    [Fact]
    public void RunFailureIsRestartable()
    {
        var lc = Sealed(VMState.Failed(VMFailurePhase.Run, "x"));
        var effects = lc.Handle(VMEvent.PowerOn);
        Assert.Equal(VMState.AwaitingPhoneUnlock, lc.State);
        Assert.Equal(new[] { VMEffect.StartVirtualMachine }, effects);
    }

    // ---- Invalid transitions are loud, not swallowed ----

    [Fact]
    public void InvalidTransitionsThrow()
    {
        var cases = new (VMState state, VMEvent ev)[]
        {
            (VMState.Created, VMEvent.PowerOn),               // nothing installed yet
            (VMState.Created, VMEvent.InstallSucceeded),
            (VMState.Installing, VMEvent.PowerOn),            // mid-install
            (VMState.Installing, VMEvent.GuestUnlocked),
            (VMState.Installed, VMEvent.StartInstall),        // already installed
            (VMState.Installed, VMEvent.GuestUnlocked),       // not booted
            (VMState.Running, VMEvent.PowerOn),               // already up
            (VMState.Running, VMEvent.GuestUnlocked),
            (VMState.Running, VMEvent.StartInstall),
            (VMState.Stopped, VMEvent.GuestUnlocked),
            (VMState.AwaitingPhoneUnlock, VMEvent.PowerOn),
            (VMState.Failed(VMFailurePhase.Install, "x"), VMEvent.PowerOn), // must retry install
            (VMState.Failed(VMFailurePhase.Run, "x"), VMEvent.StartInstall),
        };
        foreach (var (state, ev) in cases)
        {
            var lc = Sealed(state);
            var thrown = Assert.Throws<VMLifecycleException>(() => lc.Handle(ev));
            Assert.Equal(state, thrown.From);
            Assert.Equal(ev, thrown.On);
            Assert.Equal(state, lc.State); // state must not change on a rejected event
        }
    }

    // ---- Injectable clock ----

    [Fact]
    public void StateTimestampsComeFromTheInjectedClock()
    {
        var times = new Queue<DateTimeOffset>(new[]
        {
            DateTimeOffset.FromUnixTimeSeconds(100),
            DateTimeOffset.FromUnixTimeSeconds(200),
            DateTimeOffset.FromUnixTimeSeconds(300),
        });
        var lc = new VMLifecycle(sealedAtBoot: true, clock: () => times.Dequeue());
        Assert.Equal(DateTimeOffset.FromUnixTimeSeconds(100), lc.StateChangedAt);
        lc.Handle(VMEvent.StartInstall);
        Assert.Equal(DateTimeOffset.FromUnixTimeSeconds(200), lc.StateChangedAt);
        lc.Handle(VMEvent.InstallSucceeded);
        Assert.Equal(DateTimeOffset.FromUnixTimeSeconds(300), lc.StateChangedAt);
    }

    // ---- The duration-gated install verdict (the Mac Phase-0 hard-won finding) ----

    [Fact]
    public void CleanStopAfterPlausibleDurationIsSuccess()
    {
        Assert.Equal(VMEvent.InstallSucceeded, VMLifecycle.VerdictForCleanInstallStop(TimeSpan.FromMinutes(10)));
        // Edge: exactly the minimum counts as success.
        Assert.Equal(VMEvent.InstallSucceeded, VMLifecycle.VerdictForCleanInstallStop(VMLifecycle.MinPlausibleInstallDuration));
    }

    [Fact]
    public void TooFastCleanStopIsAnActionableFailure()
    {
        var verdict = VMLifecycle.VerdictForCleanInstallStop(TimeSpan.FromSeconds(12));
        Assert.Equal(VMEventKind.InstallFailed, verdict.Kind);
        Assert.Contains("12s", verdict.Reason);
        Assert.Contains("Retry", verdict.Reason);
    }

    [Fact]
    public void PrebuiltSpecializationHasShortNonzeroFloor()
    {
        Assert.Equal(VMEvent.InstallSucceeded, VMLifecycle.VerdictForCleanProvisioningStop(
            VMLifecycle.MinPlausibleSpecializationDuration, VMProvisioningMode.PrebuiltAppliance));
        Assert.Equal(VMEventKind.InstallFailed, VMLifecycle.VerdictForCleanProvisioningStop(
            TimeSpan.FromMilliseconds(300), VMProvisioningMode.PrebuiltAppliance).Kind);
        Assert.Equal(VMEventKind.InstallFailed, VMLifecycle.VerdictForCleanProvisioningStop(
            VMLifecycle.MinPlausibleSpecializationDuration, VMProvisioningMode.InstallerISO).Kind);
    }

    [Fact]
    public void InstanceVerdictDerivesElapsedFromTheStateTimestamp()
    {
        var t0 = DateTimeOffset.FromUnixTimeSeconds(1000);
        var now = t0;
        var lc = new VMLifecycle(sealedAtBoot: true, clock: () => now);
        lc.Handle(VMEvent.StartInstall); // stateChangedAt = t0

        Assert.Equal(VMEventKind.InstallFailed, lc.VerdictForCleanInstallStop(t0.AddSeconds(30)).Kind);
        Assert.Equal(VMEventKind.InstallSucceeded, lc.VerdictForCleanInstallStop(t0.AddSeconds(120)).Kind);
    }

    [Fact]
    public void InstanceVerdictOutsideInstallingIsAProgrammingError()
    {
        var lc = Sealed(VMState.Running);
        Assert.Throws<InvalidOperationException>(
            () => lc.VerdictForCleanInstallStop(DateTimeOffset.UnixEpoch));
    }
}
